src/lib/ledger_builder/ledger_builder.ml

```diff
@@ -1676,18 +1608,39 @@ end = struct
     in
     (*TODO: return an or_error here *)
     let work_to_do = work_to_do_exn t'.scan_state in
-    let pre_diffs =
-      generate_prediff logger work_to_do transactions_by_fee get_completed_work
-        ledger self partitions
+    let completed_works_seq =
+      Sequence.fold_until work_to_do ~init:Sequence.empty
+        ~f:(fun seq w ->
+          match get_completed_work w with
+          | Some cw_checked ->
+              Continue (Sequence.append seq (Sequence.singleton cw_checked))
+          | None -> Stop seq )
+        ~finish:Fn.id
+    in
+    (*Transactions in reverse order for faster removal if there is no space when creating the diff*)
+    let transactions_rev =
+      Sequence.fold transactions_by_fee ~init:Sequence.empty ~f:(fun seq t ->
+          match Ledger.apply_transaction ledger (User_command t) with
+          | Error _ ->
+              Logger.error logger
+                !"Invalid user command: %{sexp: \
+                  User_command.With_valid_signature.t} \n\
+                  %!"
```

Comment:
when can this happen? Is this that case where the transaction pool thinks something is valid, but on this particular ledger it isn't?

Reply:
Yes